### PR TITLE
obs-studio: add plugin wrapper

### LINKED ISSUE #77627: Plugin packaging for OBS
Follow up to #61482

Integrate OBS plugin better with NixOS.

It intends to fix this:

https://github.com/NixOS/nixpkgs/blob/a23a985594bf7c3eb543bdaf1e7a9c5b6ddd6099/pkgs/applications/video/obs-studio/wlrobs.nix#L1-L7

### PR DESCRIPTION
**Pull request by deviant**
Once #125261 is merged, we'll have access to obsproject/obs-studio#4067, allowing OBS to load plugins from a custom path.

This adds a new function, `wrapOBS`, which allows users to customise their OBS with plugins like so:

```nix
wrapOBS {
  plugins = with obs-studio-plugins; [
    wlrobs
    obs-multi-rtmp
    obs-gstreamer
  ];
}
```

This also cleans up and moves the currently packaged OBS plugins to their own attrset, `obs-studio-plugins`.

`obs-v4l2sink` was removed entirely, as it's [now part of upstream](https://github.com/CatxFish/obs-v4l2sink/issues/56#issuecomment-753191690).

###### Motivation for this change

Fixes #77627. Impure hacks are no longer necessary in order to use OBS with plugins! :tada:

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 06kellyjac** [comment]
Will the plugins continue to work when linked to `~/.config/obs-studio/plugins/` or do they **have** to be added with `wrapOBS`? I'll probably update the obs module in home-manager either way but it'd be good to know if this is a breaking change there

**Comment by deviant** [comment]
> Will the plugins continue to work when linked to `~/.config/obs-studio/plugins/` or do they **have** to be added with `wrapOBS`?

No, they should continue to work just fine as-is. The environment variable is considered in addition to, not instead of, other plugin locations (see the patch linked in OP).

**Comment by 06kellyjac** [comment]
Awesome, thanks for clarifying.

Also when I have a PR that's open for 5-ish+ days without review I tend to throw the marvin bot in by commenting with the following:

```
/marvin opt-in
/status needs_reviewer
```

**Comment by 06kellyjac** [comment]
Result of `nixpkgs-review pr 125308` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

File: pkgs/applications/video/obs-studio/plugins/wlrobs.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  };
+
+  nativeBuildInputs = [ meson pkg-config ninja ];
+  buildInputs = [ wayland obs-studio libX11 ];
+
+  meta = with lib; {
+    description = "An obs-studio plugin that allows you to screen capture on wlroots based wayland compositors";
+    homepage = "https://hg.sr.ht/~scoopta/wlrobs";
+    maintainers = with maintainers; [ grahamc V ];
+    license = licenses.gpl3;
```
Comment (SuperSandro2000) [inline pkgs/applications/video/obs-studio/plugins/wlrobs.nix:23]: gpl3Only or gpl3Plus?

Comment (deviant) [reply pkgs/applications/video/obs-studio/plugins/wlrobs.nix]: [GPL-3.0+](https://hg.sr.ht/~scoopta/wlrobs/browse/src/main.c?rev=4184a4a8ea7dc054c993efa16007f3a75b2c6f51#L7)

**Review by kevincox — APPROVED** [review]

**Comment by deviant** [comment]
@kevincox Thanks!

**Comment by JonathanLorimer** [comment]
@deviant is this the correct way to use `wrapOBS`?

```nix
nixpkgs.overlays = [
  (self: super: {
    obs-studio = self.wrapOBS {
      plugins = with self.obs-studio-plugins; [ wlrobs ];
    };
  })
];
```

**Comment by JonathanLorimer** [comment]
Okay, figured out that is definitely not how to do it. Looks like you just call the `wrapOBS` function where you want OBS? It would be nice to have some documentation on how to use this with home-manager. I tried this but it didn't work:
```nix
      obs-studio = {
        enable = true;
        package = pkgs.wrapOBS {
          plugins = with pkgs.obs-studio-plugins; [ wlrobs ];
        };
        plugins = with pkgs.obs-studio-plugins; [ wlrobs ];
      };
```

I am happy to write up the documentation, but am a bit unsure what the correct solution is. The relevant home-manager docs are [here](https://rycee.gitlab.io/home-manager/options.html#opt-programs.obs-studio.enable)

**Comment by deviant** [comment]
@JonathanLorimer If you want that to be what `obs-studio` refers to globally, yes. Personally I just have

```nix
{
  users.users.v.packages = with pkgs; [
    (wrapOBS {
      plugins = with obs-studio-plugins; [
        wlrobs
      ];
    })
  ];
}
```

**Comment by deviant** [comment]
Note that if you set `obs-studio` to that globally, there will probably be infinite recursion, as `wrapOBS` refers to it internally.

**Comment by JonathanLorimer** [comment]
Hmmm, my plugins directory is empty, but I think this is a home manager issue now.

**Comment by deviant** [comment]
@JonathanLorimer I don't use home-manager personally, but looking at their [OBS Studio module](https://github.com/nix-community/home-manager/blob/master/modules/programs/obs-studio.nix), this change will have broken it. It needs to be rewritten to use `wrapOBS` or have the paths set up differently (because the filesystem layout for plugins is different depending on whether they're installed in `/lib` or in `~/.config/obs-studio`, and this PR changes the layout to fit the former). This module could even be completely removed if people use `wrapOBS`, however, as it seems to just exist to allow people to install plugins.

**Comment by 06kellyjac** [comment]
I did ask if the change would break linking to `~/.config/obs-studio` :upside_down_face: 

https://github.com/NixOS/nixpkgs/pull/125308#issuecomment-858896104

**Comment by deviant** [comment]
@j-k It doesn't break any *previously linked* instances. I had some plugins already installed in `~/.config/obs-studio` myself, which continued to work just fine. I wasn't aware there was a place where this was done automatically— having home-manager out-of-tree made this easy to miss.

**Comment by deviant** [comment]
Actually, this won't even break silently b/c the previous plugin packages were added to aliases. So this is fine regardless.

**Comment by 06kellyjac** [comment]
I can set this up similar to wrapping neovim in home-manager but since there's only a packages input on the `wrapOBS` function and not a way to pass in your own `obs-studio` to be wrapped it'll either: be a breaking change removing the `package` option from the module, or risk that users set pluggins and set their own package resulting in the pluggins not actually being wrapped in.

I think I'll go with an alternative route for now, leaving it as-is and just fixing the link path but if a way to pass in your own `obs-studio` is added to `wrapOBS` then we can move it fully over

**Comment by deviant** [comment]
> I think I'll go with an alternative route for now, leaving it as-is and just fixing the link path but if a way to pass in your own `obs-studio` is added to `wrapOBS` then we can move it fully over

No, you can already do this:

```nix
wrapOBS.override { obs-studio = your-custom-obs-studio; } { plugins = [ obs-studio-plugins.wlrobs ]; }
```

**Comment by 06kellyjac** [comment]
> No, you can already do this:
> 
> ```nix
> wrapOBS.override { obs-studio = your-custom-obs-studio; } { plugins = [ obs-studio-plugins.wlrobs ]; }
> ```

Looks good, I'll do a couple more quick tests then pop in the PR

image

**Comment by 06kellyjac** [comment]
:+1: looks like the override is working too, it would have wrapped if there was actually an executable called `obs`
I don't have a custom obs that would be noticeably different so I just used `pkg.hello`

```
warning: Git tree '/home/jk/projects/personal/dotfiles-nix' is dirty
error: builder for '/nix/store/c6cs3b7slvpyjsg77ymn6v8dn7smfvgm-wrapped-hello-2.10.drv' failed with exit code 1;
       last 9 log lines:
       >
       > Builder called die: Cannot wrap '/nix/store/hyflfr3xh1b747sfqgjjh8b2xbxy9vld-wrapped-hello-2.10/bin/obs' because it is not an executable file
```

**Comment by deviant** [comment]
```
nix-repl> :b wrapOBS.override { obs-studio = pkgs.writeShellScriptBin "obs" "echo 'Hello, world!'"; } {}

this derivation produced the following outputs:
  out -> /nix/store/6w1n94x6hav4v5sqdmbqki91fdqi2xil-wrapped-obs

v@january ~> /nix/store/6w1n94x6hav4v5sqdmbqki91fdqi2xil-wrapped-obs/bin/obs
Hello, world!
```

**Comment by JonathanLorimer** [comment]
Sorry, I am a bit lost, how is the override going to fix home-manager stuff? Wouldn't this require a PR against `home-manager` to get their plugin option to use `wrapOBS`?

**Comment by 06kellyjac** [comment]
Yes, as I've mentioned in previous comments I'm writing a PR to have the home-manager module for `obs-studio` use `wrapOBS`

The screenshot demos it working

**Comment by GrabbenD** [comment]
Using it successfully with Home Manager like this on Hyprland/wlroots 😄 
```
programs.obs-studio = {
  enable = true;
  plugins = with pkgs.obs-studio-plugins; [
    wlrobs
  ];
};
```

**Pull request merged**